pyquil/api/qvm.py

```diff
@@ -142,46 +151,55 @@ def run_and_measure(self, quil_program, qubits, trials=1):
         :param Program quil_program: A Quil program.
         :param list|range qubits: A list of qubits.
         :param int trials: Number of shots to collect.
+        :param needs_compilation: If True, preprocesses the job with the compiler.
+        :param isa: If set, compiles to this target ISA.
         :return: A list of a list of bits.
         :rtype: list
         """
-        payload = self._run_and_measure_payload(quil_program, qubits, trials)
+        if needs_compilation and not self.use_queue:
+            raise TypeError("Synchronous endpoint does not support compilation preprocessing.  Use run_and_measure_async instead.")
```

Comment:
Alternatively, we could just switch to using the queue in this case.

Reply:
Up to y'all. In general I'm more comfortable alerting the user that we can't complete the task as ordered rather than quietly meeting them halfway—though in this case it is, indeed, a pretty subtle-feeling difference on the user's end.

Reply:
I would prefer to meet them halfway, we already switch on the queue for the _async methods anyway